File: gen_calendar.py
#!/usr/bin/env python3
"""Generate a 14-week calendar sheet in an Excel workbook."""

import argparse
import sys
from datetime import date, timedelta

import openpyxl
from openpyxl.styles import Font, PatternFill, Alignment


GRAY_FILL = PatternFill(patternType="solid", fgColor="F0F0F0")
MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun",
          "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]


def prev_sunday(d: date) -> date:
    """Return d if it's a Sunday, else the preceding Sunday."""
    return d - timedelta(days=d.weekday() + 1 if d.weekday() != 6 else 0)


def cell_label(d: date, prev: date | None) -> str:
    """Return 'Mon D' at a month boundary, else just 'D'."""
    if prev is None or d.month != prev.month:
        return f"{MONTHS[d.month - 1]} {d.day}"
    return str(d.day)


def _core_months(start: date, weeks: int = 14) -> list[date]:
    """Months with 14+ days in the range (trims fringe months at either end)."""
    end = start + timedelta(weeks=weeks) - timedelta(days=1)
    result = []
    cur = start.replace(day=1)
    while cur <= end:
        next_month = date(cur.year + (cur.month == 12), cur.month % 12 + 1, 1)
        days_in_range = (min(next_month - timedelta(days=1), end) -
                         max(cur, start)).days + 1
        if days_in_range >= 14:
            result.append(cur)
        cur = next_month
    return result or [start.replace(day=1)]


def sheet_title(start: date, weeks: int = 14) -> str:
    months = _core_months(start, weeks)
    if len(months) == 1:
        return months[0].strftime("%B %Y")
    first, last = months[0], months[-1]
    if first.year == last.year:
        return f"{first.strftime('%B')} - {last.strftime('%B')} {first.year}"
    return f"{first.strftime('%B %Y')} - {last.strftime('%B %Y')}"


def tab_name(start: date, weeks: int = 14) -> str:
    months = _core_months(start, weeks)
    if len(months) == 1:
        return months[0].strftime("%b %Y")
    first, last = months[0], months[-1]
    if first.year == last.year:
        return f"{first.strftime('%b')}-{last.strftime('%b')} {first.year}"
    return f"{first.strftime('%b %Y')} - {last.strftime('%b %Y')}"


def build_sheet(wb: openpyxl.Workbook, start: date, weeks: int = 14):
    name = tab_name(start, weeks)
    if name in wb.sheetnames:
        print(f"Sheet '{name}' already exists — overwriting.", file=sys.stderr)
        del wb[name]

    ws = wb.create_sheet(title=name)
    bold = Font(bold=True)
    center = Alignment(horizontal="center", vertical="center")
    left = Alignment(horizontal="left", vertical="top")

    # Row 1: merged title
    ws.merge_cells("A1:H1")
    ws["A1"] = sheet_title(start, weeks)
    ws["A1"].font = bold
    ws["A1"].alignment = center
    ws.row_dimensions[1].height = 30.75

    # Row 2: day headers
    for col, h in enumerate(["Notes", "Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"], 1):
        cell = ws.cell(row=2, column=col, value=h)
        cell.font = bold
        cell.alignment = center
    ws.row_dimensions[2].height = 15

    ws.column_dimensions["A"].width = 7.13
    ws.column_dimensions["B"].width = 8.38

    all_days = [start + timedelta(days=i) for i in range(weeks * 7)]

    for week in range(weeks):
        date_row = 3 + week * 2
        empty_row = date_row + 1
        ws.row_dimensions[date_row].height = 17.25
        ws.row_dimensions[empty_row].height = 17.25
        ws.merge_cells(f"A{date_row}:A{empty_row}")
        ws[f"A{date_row}"].alignment = center

        for day_offset in range(7):
            idx = week * 7 + day_offset
            d = all_days[idx]
            prev = all_days[idx - 1] if idx > 0 else None
            cell = ws.cell(row=date_row, column=2 + day_offset, value=cell_label(d, prev))
            cell.alignment = left
            if day_offset in (0, 6):
                cell.fill = GRAY_FILL

    return ws


def parse_date(s: str) -> date:
    try:
        return date.fromisoformat(s)
    except ValueError:
        print(f"Error: invalid date '{s}'. Use YYYY-MM-DD format.", file=sys.stderr)
        sys.exit(1)


def main():
    parser = argparse.ArgumentParser(
        description="Add a 14-week calendar sheet to an Excel workbook.",
        epilog="Example: gen_calendar.py 2026-04-05",
    )
    parser.add_argument(
        "start_date",
        help="Start date (YYYY-MM-DD). Snapped to the preceding Sunday.",
    )
    parser.add_argument(
        "--file", "-f",
        default="Calendar Templates.xlsx",
        help="Excel file path (default: 'Calendar Templates.xlsx')",
    )
    parser.add_argument(
        "--weeks", "-w",
        type=int,
        default=14,
        help="Number of weeks (default: 14)",
    )
    args = parser.parse_args()

    start = prev_sunday(parse_date(args.start_date))

    try:
        wb = openpyxl.load_workbook(args.file)
    except FileNotFoundError:
        wb = openpyxl.Workbook()
        if "Sheet" in wb.sheetnames:
            del wb["Sheet"]

    build_sheet(wb, start, args.weeks)
    wb.save(args.file)

    tab = tab_name(start, args.weeks)
    end = start + timedelta(weeks=args.weeks) - timedelta(days=1)
    print(f"Added sheet '{tab}' ({sheet_title(start, args.weeks)}) to {args.file}")
    print(f"  {start} – {end} ({args.weeks} weeks)")


if __name__ == "__main__":
    main()

File: src/bin/gen-calendar.rs
//! Generate a 14-week calendar sheet in an Excel workbook.

use chrono::{Datelike, Duration, NaiveDate};
use clap::Parser;
use rust_xlsxwriter::{Color, Format, FormatAlign, FormatBorder, Workbook};

const MONTHS: [&str; 12] = [
    "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
];
const GRAY_FILL: Color = Color::RGB(0xF0F0F0);

#[derive(Parser)]
#[command(
    about = "Create a new Excel file with a 14-week calendar sheet.",
    after_help = "Example: gen-calendar 2026-04-05"
)]
struct Args {
    /// Start date (YYYY-MM-DD). Snapped to the preceding Sunday.
    start_date: String,

    /// Output file path (default: '<tab-name>.xlsx', e.g. 'Apr-Jun 2026.xlsx')
    #[arg(short, long)]
    output: Option<String>,

    /// Number of weeks
    #[arg(short, long, default_value_t = 14)]
    weeks: u32,
}

fn parse_date(s: &str) -> NaiveDate {
    match NaiveDate::parse_from_str(s, "%Y-%m-%d") {
        Ok(d) => d,
        Err(_) => {
            eprintln!("Error: invalid date '{s}'. Use YYYY-MM-DD format.");
            std::process::exit(1);
        }
    }
}

/// Return d if it's a Sunday, else the preceding Sunday.
fn prev_sunday(d: NaiveDate) -> NaiveDate {
    let days_since_sunday = d.weekday().num_days_from_sunday();
    d - Duration::days(days_since_sunday as i64)
}

/// Return "Mon D" at a month boundary, else just "D".
fn cell_label(d: NaiveDate, prev: Option<NaiveDate>) -> String {
    match prev {
        Some(p) if p.month() == d.month() => d.day().to_string(),
        _ => format!("{} {}", MONTHS[d.month0() as usize], d.day()),
    }
}

/// Months with 14+ days in the range (trims fringe months at either end).
fn core_months(start: NaiveDate, weeks: u32) -> Vec<NaiveDate> {
    let end = start + Duration::weeks(weeks as i64) - Duration::days(1);
    let mut result = Vec::new();
    let mut cur = start.with_day(1).unwrap();
    while cur <= end {
        let next_month = if cur.month() == 12 {
            NaiveDate::from_ymd_opt(cur.year() + 1, 1, 1).unwrap()
        } else {
            NaiveDate::from_ymd_opt(cur.year(), cur.month() + 1, 1).unwrap()
        };
        let range_end = std::cmp::min(next_month - Duration::days(1), end);
        let range_start = std::cmp::max(cur, start);
        let days_in_range = (range_end - range_start).num_days() + 1;
        if days_in_range >= 14 {
            result.push(cur);
        }
        cur = next_month;
    }
    if result.is_empty() {
        result.push(start.with_day(1).unwrap());
    }
    result
}

fn sheet_title(start: NaiveDate, weeks: u32) -> String {
    let months = core_months(start, weeks);
    let (first, last) = (months[0], *months.last().unwrap());
    if months.len() == 1 {
        format!("{}", first.format("%B %Y"))
    } else if first.year() == last.year() {
        format!("{} - {} {}", first.format("%B"), last.format("%B"), first.year())
    } else {
        format!("{} - {}", first.format("%B %Y"), last.format("%B %Y"))
    }
}

fn tab_name(start: NaiveDate, weeks: u32) -> String {
    let months = core_months(start, weeks);
    let (first, last) = (months[0], *months.last().unwrap());
    if months.len() == 1 {
        format!("{}", first.format("%b %Y"))
    } else if first.year() == last.year() {
        format!("{}-{} {}", first.format("%b"), last.format("%b"), first.year())
    } else {
        format!("{} - {}", first.format("%b %Y"), last.format("%b %Y"))
    }
}

fn build_sheet(
    workbook: &mut Workbook,
    start: NaiveDate,
    weeks: u32,
) -> Result<(), rust_xlsxwriter::XlsxError> {
    let name = tab_name(start, weeks);
    let sheet = workbook.add_worksheet();
    sheet.set_name(&name)?;

    let title_format = Format::new()
        .set_bold()
        .set_align(FormatAlign::Center)
        .set_align(FormatAlign::VerticalCenter);
    let header_format = Format::new().set_bold().set_align(FormatAlign::Center);
    let week_num_format = Format::new()
        .set_align(FormatAlign::Center)
        .set_align(FormatAlign::VerticalCenter);

    let date_format = Format::new()
        .set_align(FormatAlign::Left)
        .set_align(FormatAlign::Top)
        .set_border_top(FormatBorder::Thin)
        .set_border_left(FormatBorder::Thin)
        .set_border_right(FormatBorder::Thin);
    let date_format_gray = date_format.clone().set_background_color(GRAY_FILL);

    let bottom_format = Format::new()
        .set_border_bottom(FormatBorder::Thin)
        .set_border_left(FormatBorder::Thin)
        .set_border_right(FormatBorder::Thin);
    let bottom_format_gray = bottom_format.clone().set_background_color(GRAY_FILL);

    // Row 1: merged title
    sheet.merge_range(0, 0, 0, 7, &sheet_title(start, weeks), &title_format)?;
    sheet.set_row_height(0, 30.75)?;

    // Row 2: day headers
    for (col, h) in ["Notes", "Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"]
        .iter()
        .enumerate()
    {
        sheet.write_string_with_format(1, col as u16, *h, &header_format)?;
    }
    sheet.set_row_height(1, 15)?;

    // set_column_width() bakes in extra cell padding that Excel's own
    // pixels-to-chars formula doesn't remove when writing the width back out,
    // so it overshoots the requested character width. Pass pixels directly
    // (chars * default digit width) to land close to the intended value.
    sheet.set_column_width_pixels(0, (7.13_f64 * 7.0).round() as u32)?;
    sheet.set_column_width_pixels(1, (8.38_f64 * 7.0).round() as u32)?;

    let all_days: Vec<NaiveDate> = (0..weeks * 7)
        .map(|i| start + Duration::days(i as i64))
        .collect();

    for week in 0..weeks {
        let date_row = 2 + week * 2;
        let empty_row = date_row + 1;
        sheet.set_row_height(date_row, 17.25)?;
        sheet.set_row_height(empty_row, 17.25)?;
        sheet.merge_range(date_row, 0, empty_row, 0, "", &week_num_format)?;

        for day_offset in 0..7u32 {
            let idx = (week * 7 + day_offset) as usize;
            let d = all_days[idx];
            let prev = if idx > 0 { Some(all_days[idx - 1]) } else { None };
            let col = (1 + day_offset) as u16; // column B (index 1) = Sunday
            let is_weekend = day_offset == 0 || day_offset == 6;

            let date_fmt = if is_weekend { &date_format_gray } else { &date_format };
            sheet.write_string_with_format(date_row, col, cell_label(d, prev), date_fmt)?;

            let bottom_fmt = if is_weekend { &bottom_format_gray } else { &bottom_format };
            sheet.write_blank(empty_row, col, bottom_fmt)?;
        }
    }

    Ok(())
}

fn main() {
    let args = Args::parse();

    let start = prev_sunday(parse_date(&args.start_date));

    let mut workbook = Workbook::new();
    build_sheet(&mut workbook, start, args.weeks).unwrap_or_else(|e| {
        eprintln!("Error building sheet: {e}");
        std::process::exit(1);
    });

    let tab = tab_name(start, args.weeks);
    let out = args
        .output
        .clone()
        .unwrap_or_else(|| format!("{tab} {start}.xlsx"));

    workbook.save(&out).unwrap_or_else(|e| {
        eprintln!("Error saving {out}: {e}");
        std::process::exit(1);
    });

    let end = start + Duration::weeks(args.weeks as i64) - Duration::days(1);
    println!("Created {out} — '{tab}' ({})", sheet_title(start, args.weeks));
    println!("  {start} – {end} ({} weeks)", args.weeks);
}
